modules/statistics/php/charts.class.inc

```diff
@@ -0,0 +1,343 @@
+                    $scanData['labels']
+                )
+            );
+        }
+        return (new \LORIS\Http\Response\JsonResponse($scanData));
+    }
+
+    /**
+     * Helper to get the scans per month data in the correct format for a
+     * given site
+     *
+     * @param array $data   The data for all sites.
+     * @param int   $siteID The site to get the monthly data for.
+     * @param array $labels An array of all valid labels. (Note that not
+     *                      all labels may have data in $data, in which
+     *                      case this function must include a "0" to maintain
+     *                      order.)
+     *
+     * @return int An indexed array of data for site in the order of $labels
+     */
+    private function _getScansPerMonthData(array $data, int $siteID, array $labels)
+    {
+        $sitedata   = array_filter(
+            $data,
+            function ($row) use ($siteID) {
+                return $row['CenterID'] == $siteID;
+            }
+        );
+        $mappeddata = [];
+        foreach ($sitedata as $row) {
+            $mappeddata[$row['datelabel']] = $row['count'];
+        }
+
+        $data = [];
+        foreach ($labels as $i => $label) {
+            $data[$i] = $mappeddata[$label] ?? 0;
+        }
+        return $data;
+    }
+
+    /**
+     * Helper to handle the siterecruitment_line endpoint.
+     *
+     * @return ResponseInterface
+     */
+    private function _handleSiteLineData()
+    {
+        $DB = \NDB_Factory::singleton()->database();
+
+        $recruitmentData      = array();
+        $recruitmentStartDate = $DB->pselectOne(
+            "SELECT MIN(Date_registered) FROM candidate",
+            array()
+        );
+        $recruitmentEndDate   = $DB->pselectOne(
+            "SELECT MAX(Date_registered) FROM candidate",
+            array()
+        );
+
+        $recruitmentData['labels'] = $this->_createSiteLineChartLabels(
+            $recruitmentStartDate,
+            $recruitmentEndDate
+        );
+
+        $list_of_sites = \Utility::getSiteList(true, false);
+
+        foreach ($list_of_sites as $siteID => $siteName) {
+            $recruitmentData['datasets'][] = array(
+                "name" => $siteName,
+                "data" => $this->_getSiteLineRecruitmentData(
+                    $siteID,
+                    $recruitmentData['labels']
+                ),
+            );
+        }
+        return new \LORIS\Http\Response\JsonResponse($recruitmentData);
+    }
+
+    /**
+     * Helper to generate labels for every month between startDate and endDate.
+     *
+     * @param string $startDate The start date for the labels in YYYY-MM-DD format.
+     * @param string $endDate   The end date for the labels in YYYY-MM-DD format.
+     *
+     * @return array
+     */
+    private function _createSiteLineChartLabels(string $startDate, string $endDate)
+    {
+        $startDateYear  = substr($startDate, 0, 4);
```

Comment:
I know you didn't write this code but do you think we could refactor this part?

My preference is to use DateTimes to do this but if you don't want to do that adding some class constants and removing the magic numbers would be helpful.

Reply:
Use DateTimes to do what? This is a helper to generate labels for every month between 2 dates, I don't see how DateTime objects would help/be used.

Reply:
This should do the same thing and is a lot easier to read imo.

```php
// Create example dates
$startDate = new DateTime(); // today
$endDate = $start->add(date_interval_create_from_date_string('6 months'));
$month = date_interval_create_from_date_string('1 month');
$betweenDate = clone $startDate;

while ($betweenDate < $endDate) {
    // DateTime::add modifies (increments) the $betweenDate object
    $labels[] = $betweenDate->add($month)->format('m-Y');
}
```

Reply:
I had to make a few bug fixes to make sure the results were the same, but I rewrote it to use datetime objects.